### ISSUE TITLE
Build a better cleanup

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus.DataBus.AzureBlobStorage#14]
The cleanup implemented today is constrained by the design choice how blobs are stored.
We should address #1, #2, #12, and #13 all together to address these issues.

Issue #2 (more efficient removal) - requires more efficient storage first
Issue #12 & #13 (cleanup interval & warnings) - requires having a single instance to perform cleanup rather than multiple instances running in parallel and competing over the same blobs --> (lease blob as an option)
#1 

**Update**
Plan of attack is outlined in the comments: https://github.com/Particular/NServiceBus.DataBus.AzureBlobStorage/issues/14#issuecomment-272218473

**Comment by wettsten** [comment]
It's tangentially related but a subscribing endpoint shouldn't have to configure the container and base path of where to find blobs of events to which it is subscribed.

Example:
- EndpointA and EndpointB maintain the cleaning of their own containers to increase cleanup performance (each only have to check its own published event blobs).
- EndpointA publishes EventA and uses blob container named EndpointA.
- EndpointB publishes EventB and uses blob container named EndpointB.
- EndpointC cannot subscribe to both EventA and EventB since it can only configure one container.

Possible solutions:
- Build container name and base path into the existing databus header value: {container}/{basepath}/{guid}
- Create new header keys for sending container and base path

**Comment by yvesgoeleven** [comment]
@wettsten that would cause a security problem though, to make that work we'd have to either make the blob publicly accessible or sent the connectionstring information around in the header to get access to it.

**Comment by SeanFeldman** [comment]
Here's a summary of a discussion:
1. Cleanup interval should be configurable (#12).
1. Cleanup executions should not overlap. If currently executing cleanup task is not finished, a new one should not be instantiated (related to #1).
1. Use blob lease to prevent concurrent cleanups by multiple endpoints (related to #1).
1. Use separate containers for blobs with TTL. Make removal of expired items more efficient (issue #2)*.

\* Will require a new option for enabling new mode of cleanup that will be an opt-in option.

#### Using separate containers for blobs with TTL

Blobs would be stored in separate containers for messages with TTL. Container name would be based on the cleanup frequency required (per hour/day/month). Messages w/o TTL would be stored in the default (`databus`) container and should be optionally monitored/cleanup by users as before.

**Comment by wettsten** [comment]
@yvesgoeleven i don't think there's a need to go all the way to different storage accounts. if we can assume at least a common storage account. it shouldn't be so difficult to pass container and base path in a header. that way each host would only be responsible for cleaning up the blobs that it puts in storage (if so configured).

**Comment by wettsten** [comment]
@SeanFeldman does TTL refer to the ValidUntilUtc metadata? could we configure different paths within the default container instead of different containers?

**Comment by yvesgoeleven** [comment]
@wettsten TTL does indeed refer to the ValidUntilUtc metadata, we're now thinking to split the data in containers based on ttl instead, in that way we can cleanup the container in a single shot when expired and there is no need to list each blob individually. Path itself is part of blob name, so in order to 'filter it out', we'd still need to list them all before filtering, it would still not be performant if you have a lot.

**Comment by Tsjunne** [comment]
@wettsten Wouldn't SAS tokens be a possible solution for the subscriber access? They could also contain the timetolive as expiration time if needed.

**Comment by SeanFeldman** [comment]
Another idea (credit goes to @abombss) is too handle cleanup as a deferred message.

**Comment by danielmarbach** [comment]
Related to https://github.com/Particular/PlatformDevelopment/issues/1099

**Comment by danielmarbach** [comment]
We are capturing all data bus-related issues in an internal backlog. Currently, there are no plans to rework the data bus but we might come back to the data bus in the future. Closing this for now.
